iceoryx_posh/test/integrationtests/test_service_discovery.cpp

```diff
@@ -107,615 +161,927 @@ class ServiceDiscoveryPubSub_test : public RouDi_GTest
-                     0U,
-                     iox::popo::createNotificationCallback(testCallback))
-        .and_then([]() { GTEST_SUCCEED(); })
-        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+    this->testFindService({"a"}, {"d"}, {"c"});
 }
 
-TEST_F(ServiceDiscoveryPubSub_test, ServiceDiscoveryIsNotifiedbyWaitSetAboutSingleService)
+TYPED_TEST(ServiceDiscoveryFindService_test, RepeatedSearchYieldsSameResult)
 {
-    ::testing::Test::RecordProperty("TEST_ID", "f1cf36b5-3db2-4e6f-8e05-e7e449530ec0");
-    iox::popo::WaitSet<1U> waitSet;
+    ::testing::Test::RecordProperty("TEST_ID", "f1ee22fb-dc40-4011-b054-35a31d7ee5ee");
+    this->add({"a", "b", "c"});
+    this->add({"a", "b", "aa"});
+    this->add({"aa", "a", "c"});
+    this->add({"a", "ab", "a"});
 
-    waitSet
-        .attachEvent(this->sut,
-                     ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED,
-                     0U,
-                     iox::popo::createNotificationCallback(testCallback))
-        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+    this->testFindService({"a"}, {"b"}, {"aa"});
+    auto previousResult = serviceContainer;
 
-    const iox::capro::ServiceDescription SERVICE_DESCRIPTION("Moep", "Fluepp", "Shoezzel");
-    iox::popo::UntypedPublisher publisher(SERVICE_DESCRIPTION);
+    this->testFindService({"a"}, {"b"}, {"aa"});
+    sortAndCompare(previousResult, serviceContainer);
+}
 
-    auto notificationVector = waitSet.timedWait(1_s);
+TYPED_TEST(ServiceDiscoveryFindService_test, FindNonExistingService)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "6f953d0d-bae3-45a1-82e7-c78a32b6d365");
+    this->add({"a", "b", "c"});
+
+    // those are all representatives of equivalence classes of mismatches
+    // that hould not be found
+    this->testFindService({"x"}, {"b"}, {"c"});
+    this->testFindService({"a"}, {"x"}, {"c"});
+    this->testFindService({"a"}, {"b"}, {"x"});
+    this->testFindService({"x"}, {"x"}, {"c"});
+    this->testFindService({"a"}, {"x"}, {"x"});
+    this->testFindService({"x"}, {"b"}, {"x"});
+    this->testFindService({"x"}, {"x"}, {"x"});
+}
 
-    for (auto& notification : notificationVector)
-    {
-        (*notification)();
-    }
+TYPED_TEST(ServiceDiscoveryFindService_test, FindNonExistingServiceAmongMultipleNearMatches)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "86b4977f-53c6-495c-add3-524c9f3e0f27");
+    // add multiple near matches
+    this->add({"x", "b", "c"});
+    this->add({"a", "x", "c"});
+    this->add({"a", "b", "x"});
+    this->add({"x", "x", "c"});
+    this->add({"a", "x", "x"});
+    this->add({"x", "b", "x"});
+    this->add({"x", "x", "x"});
+
+    this->testFindService({"a"}, {"b"}, {"c"});
+}
 
-    EXPECT_TRUE(callbackWasCalled);
+TYPED_TEST(ServiceDiscoveryFindService_test, FindWhenSameInstanceWithDifferentServiceAndEventExists)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "");
```

Comment:
This test and the one below can be removed since these cases are already tested in the test above.